ARSN-485: Prefix KMS Keys with an arn (HF cldsrv 9.2.0.12)

BourgoisMickael opened the pull request:
To keep track of backend provider

For AWS use key Arn instead of KeyId by default
to allow for cross-account

___

See https://scality.atlassian.net/browse/S3C-9949
This hotfix to handle SSE with both internal and external KMS. Needed to built a custom image of cloudserver & vault.

Cherry-pick for vault: https://github.com/scality/Arsenal/pull/2350

It will then be cherry-picked into latest development for 9.5

BourgoisMickael committed:
ARSN-485: Prefix KMS Keys with an arn

To keep track of backend provider

For AWS use key Arn optionally instead of KeyId by default
to allow for cross-account

Helper functions to validate and extract provider
from new arnPrefix

bert-e commented:
# Hello bourgoismickael,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/arsenal/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-485 contains:


* *None*



Considering where you are trying to merge, I expected to find at least:



* `7.70.4.1`


Please check the `Fix Version/s` of ARSN-485, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

BourgoisMickael commented on lib/network/kmsAWS/Client.ts:
```diff
@@ -101,20 +114,24 @@ export default class Client {
             logger.debug("AWS KMS: master encryption key created", { KeyMetadata: data?.KeyMetadata });
-            cb(null, data?.KeyMetadata?.KeyId);
+            const keyId = this.noAwsArn ? data?.KeyMetadata?.KeyId : data?.KeyMetadata?.Arn ?? data?.KeyMetadata?.KeyId;
+            const arn = `${this.arnPrefix}${keyId}`;
```
Yes, this is expected, I've added a comment here to describe the use case and workaround

BourgoisMickael committed:
ARSN-485: Fix broken CI ubuntu24 hotfix branch

Bump actions form v2 to v4 as setup-node has an error
```
/usr/local/bin/yarn cache dir
/home/runner/.cache/yarn/v6
Error: Cache service responded with 422
```

BourgoisMickael committed:
ARSN-485: Test functional KMS arn

BourgoisMickael commented on lib/network/KMSInterface.ts:
```diff
@@ -0,0 +1,68 @@
+/**
+ * Possible arn are:
+ * 
+ * - `arn:scality:kms:int:scality:safenet:key/$keyId` (deprecated provider)
```
I consider the deprecated `Safenet / Gemalto/Thales` using the `scality-kms` backend as internal provider with a protocol = `scality` and provider name = `safenet`.

Is this ok ?

BourgoisMickael replied:
After discussion with david, this one should rather be an external

BourgoisMickael replied:
Fixed https://github.com/scality/Arsenal/pull/2349/files#diff-1c24d37a19bc08405b548ef9b5ac12a2ef83b7edb3a2ee62cc8426c7df8e5b95R23-R26

jonathan-gramain commented on lib/network/KMSInterface.ts:
```diff
@@ -0,0 +1,68 @@
+export type AwsLikeKmsArn<T extends KmsType = KmsType> = `${AwsLikeKmsArnPrefix<T>}${string}`;
+
+/**
+ * Drop scality arn prefix from a KMS Key so it can be sent to be backend KMS
```
```suggestion
 * Drop scality arn prefix from a KMS Key so it can be sent to the backend KMS
```

BourgoisMickael replied:
Fixed https://github.com/scality/Arsenal/pull/2349/files#diff-1c24d37a19bc08405b548ef9b5ac12a2ef83b7edb3a2ee62cc8426c7df8e5b95R75

jonathan-gramain commented on lib/network/kmsAWS/Client.ts:
```diff
@@ -101,20 +114,27 @@ export default class Client {
             }
             logger.debug("AWS KMS: master encryption key created", { KeyMetadata: data?.KeyMetadata });
-            cb(null, data?.KeyMetadata?.KeyId);
+            const keyId = this.noAwsArn ? data?.KeyMetadata?.KeyId : data?.KeyMetadata?.Arn ?? data?.KeyMetadata?.KeyId;
```
This looks quite cryptic, could you spread it on multiple lines with explicit `if`s instead?

BourgoisMickael replied:
```javascript
            const keyMd = data?.KeyMetadata;

            const keyId = this.noAwsArn
                ? keyMd?.KeyId
                : keyMd?.Arn || keyMd?.KeyId;

```
Does this looks better ?

nicolas2bert replied:
with if/else? Could it be more readable?

 ```
const keyMetadata = data?.KeyMetadata;

let keyId;

if (this.noAwsArn) {
    // Use KeyId when ARN is not required
    keyId = keyMetadata?.KeyId;
} else {
    // Prefer ARN, but fall back to KeyId if ARN is missing
    keyId = keyMetadata?.Arn ?? keyMetadata?.KeyId;
}
```

BourgoisMickael replied:
Fixed https://github.com/scality/Arsenal/pull/2349/files#diff-fb4f257ae2f25bd280160e3cdec22c57fdd9aaa0ee35637e056146ce2b35e6ecR124-R131

jonathan-gramain commented on lib/network/kmsAWS/Client.ts:
```diff
@@ -23,6 +26,11 @@ interface KMSOptions {
+     * Do not use AWS KMS Key arn but id (to keep backward compatibility)
+     * Not set or false by default to use arn to be future-proof (multi account, region)
+     */
+    noAwsArn?: boolean;
```
Wondering:
- can existing customers use the ARN when they already have KMS-encrypted keys or internal keys?
- if not, how do we ensure they will not run into issues? Can we enforce the correct configuration automatically in such case?

BourgoisMickael replied:
- for internal keys, they will never reach this kmsAws client. Existing key or non kmsAws key will be sent to the correct client by cloudserver that parse our own scality arn prefix  (different from the kmsAws client's aws arn) to know the key provider.

- if they already have kmsAws keys (using only the KeyId) it's not a problem, for aws KeyId and KeyArn are interchangeable (except for cross-acount). We should still verify with custom if they do handle KeyArn as AWS does.

- Correct configuration would be to use Arn by default, except if customer's own kmsAws doesn't handle Arn. This case should be written in documentation. The only difference for now is we don't plan to migrate aws KeyId to include the aws KeyArn

nicolas2bert replied:
Can you provide an example of what a keyId looks like with noAwsArn set to false or true.

`arn:scality:kms:ext:aws_kms:custom:key/arn:aws:kms:region:accountId:key/<KEY_ID>` vs 
`arn:scality:kms:ext:aws_kms:custom:key/<KEY_ID>`

BourgoisMickael replied:
Fixed https://github.com/scality/Arsenal/pull/2349/files#diff-fb4f257ae2f25bd280160e3cdec22c57fdd9aaa0ee35637e056146ce2b35e6ecR29-R39

nicolas2bert commented on lib/network/KMSInterface.ts:
```diff
@@ -0,0 +1,68 @@
+        // Do not use full arnPrefix to allow for different providerName
+        return kmsKeyIdOrArn.substring(kmsKeyIdOrArn.indexOf('/') + 1)
+    }
+    return kmsKeyIdOrArn;
```
Is the condition introduced to handle old kmsKeyIdOrArn? Can you add it in a comment?

BourgoisMickael replied:
Yes, updated the doc: https://github.com/scality/Arsenal/pull/2349/files#diff-1c24d37a19bc08405b548ef9b5ac12a2ef83b7edb3a2ee62cc8426c7df8e5b95R75-R78

nicolas2bert commented on lib/network/kmip/Client.ts:
```diff
@@ -384,19 +391,20 @@ export default class Client {
                 return this._activateBucketKey(uniqueIdentifier, logger, cb);
             }
-            return cb(null, uniqueIdentifier);
+            return cb(null, uniqueIdentifier, `${this.arnPrefix}${uniqueIdentifier}`);
```
Why do we still need to pass the `uniqueIdentifier`? 
`${this.arnPrefix}${uniqueIdentifier}` is not enough?

BourgoisMickael replied:
I've added another parameter on each KMS client to keep backwards compatibility and avoid breaking anything, but this is not used anymore.

The KMS wrapper will always pick the 3rd arg with arnPrefix. I could remove the 2nd arg without arnPrefix if needed

nicolas2bert commented on lib/network/KMSInterface.ts:
```diff
@@ -0,0 +1,68 @@
+ * @returns KeyId extracted from scality arn (for AWS can still be an aws arn)
+ */
+export function getKeyIdFromArn(kmsKeyIdOrArn: AwsLikeKmsArnPrefix | string) {
+    if (kmsKeyIdOrArn.startsWith('arn:scality:kms:')) {
```
NIT:

we could move the prefix to a constant since it is used later on.
We could also move this simple logic to a function to explain what we are doing.
exemple:
`isScalityKmsArn(...)`
// The Scality KMS ARN prefix identifying the new Scality KMS ARN format.

BourgoisMickael replied:
Added there: https://github.com/scality/Arsenal/pull/2349/files#diff-1c24d37a19bc08405b548ef9b5ac12a2ef83b7edb3a2ee62cc8426c7df8e5b95R70-R72

nicolas2bert commented on tests/functional/kmsAWS/highlevel.spec.js:
```diff
@@ -56,9 +57,58 @@ describe('KmsAWSClient', () => {
+        });
+    });
+
+    it('should create a new master encryption key with aws arn', done => {
```
Can we add a test when the KeyMetadata.Arn is missing from the response?

BourgoisMickael replied:
There was already a test above doing that, as the flag `noAwsArn` is falsy by default and the Arn was no provided, but I added a test by passing tha flag `noAwsArn` to false explicitly: https://github.com/scality/Arsenal/pull/2349/files#diff-beda77b39ab0827a93fdfc3954bd6e3232f38f334dec6851b3699262214ca633R73-R98

BourgoisMickael committed:
ARSN-485: Fix UploadPartCopy SSE

SSE from MPU overview is passed as input
insted of using bucket SSE

BourgoisMickael committed:
ARSN-485: Test unit KMSInterface helper functions

BourgoisMickael committed:
ARSN-485: Add KMS arnPrefix to inMem vault auth

BourgoisMickael committed:
ARSN-485: Bump version

dvasilas commented on lib/network/kmip/Client.ts:
```diff
@@ -302,11 +308,12 @@ export default class Client {
-     * @param keyIdentifier - The bucket key Id
+     * @param keyIdentifierOrArn - The bucket key Id
      * @param logger - Werelog logger object
      * @param cb - The callback(err: Error)
```
minor: Maybe also update `callback(err: Error)`.

Same in `createBucketKey`.

BourgoisMickael replied:
Fixed

BourgoisMickael commented:
/approve

anurag4DSB commented:
/bypass_incompatible_branch
